### PR TITLE
update and test EarlyStopping; add OptunaPruning

### PR DESCRIPTION
**Pull request by TZTsai**
## Description
- Added a callback for Optuna early pruning. 
- Updated model validation and early stopping. 
- Added a test of early stopping for TabDDPM plugin.

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/vanderschaarlab/.github/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/vanderschaarlab/.github/blob/main/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [van der Schaar Lab Styleguide](https://github.com/vanderschaarlab/.github/blob/main/STYLEGUIDE.md)
- [x] I have labelled this PR with the relevant [Type labels](https://github.com/vanderschaarlab/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -32,6 +32,7 @@ package_dir =
 python_requires = >=3.7
 
 install_requires =
+    ipython<8.0
```
Comment (robsdavis) [inline setup.cfg:35]: Why is this required? I would prefer not to pin to old versions of any more libraries if at all possible.

Comment (TZTsai) [reply setup.cfg:35]: There was an error during the test caused by the newest IPython version. Therefore I added this requirement.

Comment (DrShushen) [reply setup.cfg]: I have also seen some recent incompatibility issues with ipython, makes sense

Comment (TZTsai) [reply setup.cfg]: Maybe in the new release of IPython 8.13.2 this issue was resolved.

Comment (TZTsai) [reply setup.cfg]: Thanks!
